### ISSUE TITLE
large datasets - how to handle them?

### ISSUE THREAD

**Issue by nort3x** [Kotlin/dataframe#141]
i read docs couldn't really find anything clarifying this, although title in readme file pretty much says "in-memory"

suppose i have a 6gb structured dataset,
in pandas for instance i can specify chunksize what is equivalent idiomatic way in dataframe? ( i see readLine parameter in read methods but I'm not sure)

- thanks

**Comment by koperagen** [comment]
Hi! I think nobody tried it before, so there is no idiomatic way to do it right now. What's the format of your dataset, CSV?

**Comment by nort3x** [comment]
### part 1: description of the problem

in my case i'm using a postgres database as my initial dataset, i first saved result of a huge hierarchical query in disk using `json_agg` and  recursive query to end up with a huge json file which is saved to file using `copy ... to` syntax 

i even later wrote this naive implementation so if i could find a way to stream data i can directly parse from sql resultset:
```kotlin
fun ResultSet.toDataFrame(): DataFrame<*>{
    val cols = with(metaData) {
        (1 .. columnCount)
            .map { getColumnName(it) }
            .toList()
    }

    val resultList = sequence {
        while (this@toDataFrame.next()) {
            val mapRepoOfObject = mutableMapOf<String,Any>()
            cols.forEach() {
                mapRepoOfObject[it] = getObject(it)
            }
            yield(mapRepoOfObject)
        }
    }.toList()

    return dataFrameOf(cols){ header ->
        resultList.map { it[header]!! }.toList()
    }
}

// use case

val con: Connection =  DriverManager.getConnection("jdbc:postgresql://blahblah/")!!
val df = con.createStatement().executeQuery("select * from huge").toDataFrame().also { println(it) }
```  

if i use limit and pagination i can produce stream of dataframes,
with the same procedure i can also end up with the same stream of dataframes from file


### part 2: so what's the problem?
for instance have groupby or any aggregate in mind, processing multiple dataframes in separate fashion can result in duplicates?

so basically i'm inspecting if we can use DataFrame for larger datasets some how

**Comment by koperagen** [comment]
Hello again. I've found some interesting article on groupBy / aggregate for chunked dataframes in pandas https://maxhalford.github.io/blog/pandas-streaming-groupby/. The general idea is that if you need a groping by some characteristic, sorting a dataset by this characteristic could help to process your data in chunks and then merge the result

**Comment by corlaez** [comment]
I have the same concern about large dataset but a different scenario and desired solution.

I am not too interested in manipulating the dataframe as a whole but rather reading row by row (step by step), in a lazy fashion (again because my data is huge files say 8GB, 20 GB). The formats I am interested in are XML, CSV

**Comment by nort3x** [comment]
@corlaez some operations by nature require traverse of whole dataset (like transformation, count , etc ...) and some operations even require auxiliary storage (like grouping, distinct, ... )
the answer i've found is  dataframe is for researching and testing on a small subset of data
for anything extra we should use the specific tool made and optimized to be used as a pipeline on big data (like sparks and hadoop (i'm not qualified to give advice on this ))
as the result i like dataframe as is and i don't expect it to grow complex to support such usecases
although it would be very beneficial if dataframe team decide to publish extra modules to interpolate with those solutions to have a seamless experience.

**Comment by zaleslaw** [comment]
It was also asked here, on the Reddit https://www.reddit.com/r/Kotlin/s/ctNLZEq2gI

Also it should be provided a documentation page for user with different option and ways, how to handle large CSV files, what could be fine-tuned

Benchmark for some synthetic CSV is also worth to add

**Comment by kpeck-north** [comment]
Running on Android and iOS via KMP / CMP we have a similar issue with large files. Server wanted to send the file to us in arrow format but there is no iOS / Android support for that so they send it to us in JSON.

Currently the mobile side does not care about every JSON key / value so we process it as a channel and filter each record down to an object that just has the fields we want then we pass this massive chunk of JSON to the DataFrame to process. It is working OK now but would be nice to not have the intermediate JSON memory block.

We can also read the data from the server and put it directly into a file which helps on memory at first but it appears when we pass that file name to DataFrame it will read in the entire file into memory then convert it to its internal format. We run out of memory.

Optimal solution: DataFrame to read from file with a list of keys to actually process (or ones to ignore)
Part 2 of Optimal: DataFrame to read from file in chunks instead of full load to memory then process

Something that may work shorter term: We read from file in chunks and parse it to out shortened JSON format with just the keys we want and write that directly to file instead of keeping all records in a list.

Love the query aspect of DataFrame, having a bit of a rough go with memory usage on mobile.

**Comment by Jolanrensen** [comment]
@kpeck-north I think in this case it may be better to let the server send the data in smaller JSON chunks or as [JSONL](https://jsonlines.org/). This would also save network traffic for the user :) There's no need sending JSON keys the mobile side doesn't care about, right?

The problem with JSON is that it's often one big structure. DataFrame parses it recursively from the root downwards, matching up objects by filling missing values with nulls and so on. There are multiple ways to split up a JSON file, and it often depends on the data in it. (For instance, the root could either be an array or an object).

So, I think in this case, the removal of keys or splitting up the JSON should be done before parsing it into a dataframe.

Another small piece of advice: what does your JSON look like? does it contain arrays with lots of different types of objects inside, for instance? This could result in an explosion of null-filled columns, as Kotlin has no notion of `undefined`, just `null`.
If this is the case, have a look at https://kotlin.github.io/dataframe/read.html#specify-key-value-paths. This might save a lot of memory.

Additionally, if your data has type clashes (so arrays/values/objects being mixed under the same key or in the same array), have a look at: https://kotlin.github.io/dataframe/read.html#manage-type-clashes

**Comment by kpeck-north** [comment]
We are trying to get server team to trim this down. The current data is a simple array of one object type, no arrays within the data type so basically as simple as it gets, just a lot of it.

JSON is not terse, would love to use Arrow format, the server supports it, but iOS does not. Right now I do read the JSON via a channel and only keep the data we need to run the DataFrame calculations and that is what I pass to DataFrame.
